app/agent/runtime/qemu/qemu.service.test.ts

```diff
@@ -0,0 +1,387 @@
+        void execSshCmd({ ssh }, command).catch((_err) => void 0);
+        await sleep(100);
+        return info;
+      },
+    );
+    expect(vmInfo).not.toBeNull();
+    expect(vmInfo?.status).toBe("on");
+    expect(vmInfo?.info.instanceId).toBe(runId);
+
+    // Ensure the VM was shut down
+    await expect(instance.getRuntimeInfo()).resolves.toEqual(null);
+    await expect(
+      doesFileExist(path.join("/proc", (vmInfo?.info.pid as number).toString())),
+    ).resolves.toEqual(false);
+  })(),
+);
+
+test.concurrent.each(testCases)(
+  `Doesn't hang forever when ssh is unresponsive on %s`,
+  async (_name, remoteTag) =>
+    provideBlockRegistry(async ({ injector, runId, brService }) => {
+      const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+      const osCt = await brService.createContainer(osIm, "osCt");
+      const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
+      const instance = injector.resolve(Token.QemuService)(runId);
+
+      let sshCommandExitedWithin: number | null = null;
+      let sshDone: (value: unknown) => void;
+      let sshDonePromise = new Promise((resolve) => {
+        sshDone = resolve;
+      });
+      await expect(
+        instance.withRuntime(
+          {
+            ssh: {
+              username: "root",
+              password: "root",
+            },
+            vcpuCount: 1,
+            memSizeMib: 128,
+            fs: { "/": osB },
+            cleanupAfterStop: true,
+            shouldPoweroff: true,
+          },
+          async ({ ssh, runtimePid }) => {
+            const now = Date.now();
+            await waitForPromises([
+              execSshCmd({ ssh }, ["sh", "-c", "sleep 1; echo 'a'"]).catch(() => {
+                sshCommandExitedWithin = Date.now() - now;
+                sshDone(void 0);
+              }),
+              (async () => {
+                await sleep(500);
+                process.kill(runtimePid, "SIGKILL");
+              })(),
+            ]);
+          },
+        ),
+      ).rejects.toThrow();
+      await sshDonePromise;
+      expect(sshCommandExitedWithin).not.toBe(null);
+      expect(sshCommandExitedWithin).toBeLessThan(5000);
+    })(),
+);
+
+// Run those tests only on alpine, this test does not depend on the guest distro but on the guest kernel
+test.concurrent.each(
+  [testCases[0]].flatMap(([name, remoteTag]) => [
+    [1, 128, name, remoteTag],
+    [1, 256, name, remoteTag],
+    [2, 128, name, remoteTag],
+    [2, 256, name, remoteTag],
+  ]),
+)(
+  `Creates vm with %d cores and %d MiB ram on %s`,
+  async (vcpuCount, memSizeMib, _name, remoteTag) =>
+    provideBlockRegistry(async ({ injector, runId, brService }) => {
+      const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+      const osCt = await brService.createContainer(osIm, "osCt");
+      const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
```

Comment:
hide?

Reply:
It's handled globally: https://github.com/hocus-dev/hocus/pull/93/files#diff-6728b317769e3b07f2f0926e1b27c3dadcd3958b1eb186a33b3cc8b3234f6047R162